File: Array&Hashing/1488. Avoid Flood in The City/solution.py
from sortedcontainers import SortedList
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        n = len(rains)

        dryDay = SortedList()
        rainDay = defaultdict(lambda: -1)
        res = [-1] * n
        for i in range(n):
            if rains[i] == 0:
                dryDay.add(i)
            else:
                if rainDay[rains[i]] == -1: # let it rain
                    rainDay[rains[i]] = i
                else: # must dry before rain
                    idx = dryDay.bisect_right(rainDay[rains[i]])
                    if idx == len(dryDay): return []

                    res[dryDay[idx]] = rains[i]
                    dryDay.pop(idx)
                    rainDay[rains[i]] = i

        while dryDay:
            idx = dryDay.pop()
            res[idx] = 1
        return res

File: Array&Hashing/1488. Avoid Flood in The City/solution.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        int n = rains.size();
        vector<int> ans(n, -1);
        
        vector<int> dry;
        unordered_map<int,int> rain;
        for (int i=0; i<n; i++) {
            if (rains[i] == 0) {
                dry.push_back(i);
                continue;
            }

            if (rain.find(rains[i]) == rain.end()) {
                rain[rains[i]] = i;
            } else {
                int day = rain[rains[i]];
                auto it = upper_bound(dry.begin(), dry.end(), day);
                if (it == dry.end()) return {};

                ans[*it] = rains[i];
                dry.erase(it);
                rain[rains[i]] = i;
            }
        }

        for (auto& idx : dry) {
            ans[idx] = 1;
        }

        return ans;
    }
};